### PR TITLE
BUG: trace frames with numpy scalar -> ndarray functions

### LINKED ISSUE #112951: dynamo does not start tracing numpy after a graph break unless it sees an array/tensor before
### 🐛 Describe the bug

Dynamo does not start tracing a frame with numpy operations unless either 1) there is something from numpy in the previous frame, or 2) the inputs to the numpy function at the start of the frame are arrays or tensors. 
If the numpy thing takes scalars and return arrays, it is not traced.

Consider the repro from https://github.com/pytorch/pytorch/issues/109604 --- with the first array operation commented out:

```
$ cat obj_arr.py 
import torch
import numpy as np

@torch.compile
def fn(x):
    # x = x ** 2   # <<< uncomment to trace np.arange(x) below
    print("HI")
    return 2 * np.arange(x)

fn(8)
```

produces

```
$ TORCH_LOGS="+dynamo" python obj_arr.py 
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping helper /home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/contextlib.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/contextlib.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/contextlib.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping backend_cache_wrapper /home/ev-br/repos/pytorch/torch/_dynamo/eval_frame.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping _maybe_init_guarded_backend_cache /home/ev-br/repos/pytorch/torch/_dynamo/eval_frame.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping innermost_fn /home/ev-br/repos/pytorch/torch/_dynamo/eval_frame.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping _set_current_backend /home/ev-br/repos/pytorch/torch/_dynamo/eval_frame.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/contextlib.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/contextlib.py
[2023-11-04 14:34:49,690] torch._dynamo.eval_frame: [DEBUG] skipping enable_dynamic /home/ev-br/repos/pytorch/torch/_dynamo/eval_frame.py
[2023-11-04 14:34:49,691] [0/0] torch._dynamo.symbolic_convert: [INFO] Step 1: torchdynamo start tracing fn obj_arr.py:5
[2023-11-04 14:34:49,693] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line obj_arr.py:5 in fn ()
[2023-11-04 14:34:49,693] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]     @torch.compile
[2023-11-04 14:34:49,693] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line obj_arr.py:8 in fn (fn)
[2023-11-04 14:34:49,693] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]         print("HI")
[2023-11-04 14:34:49,693] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_GLOBAL print []
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_CONST HI [BuiltinVariable(print)]
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE CALL_FUNCTION 1 [BuiltinVariable(print), ConstantVariable(str)]
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.symbolic_convert: [DEBUG] break_graph_if_unsupported triggered compile
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: call_function BuiltinVariable(print) [ConstantVariable(str)] {} from user code at:
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "obj_arr.py", line 8, in fn
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     print("HI")
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] 
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.output_graph: [DEBUG] restore_graphstate: removed 0 nodes
[2023-11-04 14:34:49,694] [0/0] torch._dynamo.output_graph: [DEBUG] COMPILING GRAPH due to GraphCompileReason(reason='call_function BuiltinVariable(print) [ConstantVariable(str)] {}', user_stack=[<FrameSummary file obj_arr.py, line 8 in fn>], graph_break=True)
[2023-11-04 14:34:49,734] [0/0] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2023-11-04 14:34:49,734] [0/0] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:375 in init_ambient_guards
[2023-11-04 14:34:49,734] [0/0] torch._dynamo.guards.__guards: [DEBUG] ___skip_backend_check() or ___current_backend() == ___lookup_backend(140351235318112)  # _dynamo/output_graph.py:381 in init_ambient_guards
HI
[2023-11-04 14:34:49,754] torch._dynamo.convert_frame: [DEBUG] skipping because no torch.* resume_in_fn             obj_arr.py 8
[2023-11-04 14:34:49,855] torch._dynamo.utils: [INFO] TorchDynamo compilation metrics:
[2023-11-04 14:34:49,855] torch._dynamo.utils: [INFO] Function                           Runtimes (s)
[2023-11-04 14:34:49,855] torch._dynamo.utils: [INFO] -------------------------------  --------------
[2023-11-04 14:34:49,855] torch._dynamo.utils: [INFO] _compile.<locals>.compile_inner          0.0438

```

This is similar to https://github.com/pytorch/pytorch/issues/109604, meaning that the fix in https://github.com/pytorch/pytorch/pull/110512 is incomplete. 

### Versions

main

cc @mruberry @rgommers @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by ev-br**
Fixes #112951

Make dynamo detect that `np.arange(3)` returns a FakeTensor, so the frame needs to be traced.

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/convert_frame.py [hunk 1]
```diff
@@ -178,6 +178,18 @@ def has_tensor_in_frame(frame):
...
+    # check if there's something from numpy. XXX: refactor
+    if np and config.trace_numpy:
+        if np in frame.f_globals.values():
+            return True
+        # Detect `from numpy import arange` etc
+        from .variables.misc import get_np_to_tnp_map
+
+        for co_name in frame.f_code.co_names:
+            if co_name in frame.f_globals:
+                if id(frame.f_globals[co_name]) in {id(x) for x in get_np_to_tnp_map()}:
```
Comment (Skylion007) [inline torch/_dynamo/convert_frame.py:190]: Can `{id(x) for x in get_np_to_tnp_map()}` to change at each iteration? Or should should this set be cached?

Comment (ev-br) [reply torch/_dynamo/convert_frame.py:190]: I definitely should and will be cached! The PR needs a tweak anyway, because dynamo traces too much with `np in frame.f_globals.values()`, apparently.

**Review by lezcano — COMMENTED** [review]
Alas, this is not the fix we want. Interestingly enough, it's not too incorrect, and all those errors are probably valid (I reckon that many of them coming from the 0D / np scalar divergence tho).

The reason why we don't want to tag NumPy functions are `allowed` is because `is_allowed` marks what can go in the graph, and we decided very early that we wouldn't want to put NumPy functions in the graph, as we want the graph to keep the invariant that the FX graph just has tensors (and symbolically traced ints).

It would be possible to add a fix similar to this if we swapped the NumPy functions by our wrapped `torch._numpy` functions that go from tensors to tensors tho.

The other fix would be to split this check into its own function, and use it in a selected number of places where `is_allowed` is really used to check whether we should continue tracing or not.

**Comment by lezcano** [comment]
This fix LGTM. Now it's a matter of fixing the 0-d issues that this has uncovered

**Comment by lezcano** [comment]
(0D or otherwise)

File: test/torch_np/numpy_tests/core/test_multiarray.py [hunk 2]
```diff
@@ -1707,7 +1711,7 @@ def test_sort_degraded(self):
         assert_equal(np.sort(d), do)
         assert_equal(d[np.argsort(d)], do)
 
-    @xpassIfTorchDynamo  # (reason="order='F'")
+    @xfail  # (reason="order='F'")
```
Comment (lezcano) [inline test/torch_np/numpy_tests/core/test_multiarray.py:1714]: why does this one fail? order="F" should break properly, no? Is it accessing `.flags` that yells at us? If so, fair enough, and we should break on that or implement them

File: test/torch_np/numpy_tests/core/test_scalarmath.py [hunk 3]
```diff
@@ -129,6 +130,7 @@ def test_type_create(self):
             b = atype([1, 2, 3])
             assert_equal(a, b)
 
+    @skipIfTorchDynamo  # freezes under torch.Dynamo (loop unrolling, huh)
```
Comment (lezcano) [inline test/torch_np/numpy_tests/core/test_scalarmath.py:133]: Uf. It's not that it freezes, but it'll take a few days to run lol

File: torch/_numpy/_reductions_impl.py [hunk 4]
```diff
@@ -421,9 +421,12 @@ def percentile(
     *,
     interpolation: NotImplementedType = None,
 ):
+    dt = _dtypes_impl.default_dtypes().float_dtype
+    qq = q / torch.as_tensor(100.0, dtype=dt)
```
Comment (lezcano) [inline torch/_numpy/_reductions_impl.py:425]: How is this preferable?

Comment (ev-br) [reply torch/_numpy/_reductions_impl.py:425]: Int tensor / 100.0 is float32 otherwise. Found via an assert_equal in TestMedian failing with abs errors of the order of 1e-7.

File: test/torch_np/numpy_tests/core/test_multiarray.py (hunk 2 above)
Comment (ev-br) [reply test/torch_np/numpy_tests/core/test_multiarray.py:1714]: I plan to take a look at .flags in a follow-up.

File: torch/_numpy/_reductions_impl.py (hunk 4 above)
Comment (lezcano) [reply torch/_numpy/_reductions_impl.py:425]: Ugh, right. In order to keep the scalars being scalar for now, rather than doing this, let's do a cast for `q` when `q` is integer.

Wrapping scalars into tensors is best avoided for now, as it deactivates a number of optimisations within PyTorch.

Comment (ev-br) [reply torch/_numpy/_reductions_impl.py:425]: `q` is normalized as ArrayLike already, so it's a tensor at this point even if the original input was a scalar.

Comment (lezcano) [reply torch/_numpy/_reductions_impl.py:425]: I mean wrapping 100.0 into a tensor

Comment (ev-br) [reply torch/_numpy/_reductions_impl.py]: OK, done in https://github.com/pytorch/pytorch/pull/112959/commits/ff290b04fa1f58eb1efdd435f41317d98569d25f

Comment (lezcano) [reply torch/_numpy/_reductions_impl.py]: The current implementation still wraps 100. I meant
```python
    if _dtypes_impl.python_type_for_torch(q.dtype) == int:
        q = q.to(_dtypes_impl.default_dtypes().float_dtype)
    qq = q / 100.0
```

Comment (ev-br) [reply torch/_numpy/_reductions_impl.py]: OK, done.

**Review by lezcano — APPROVED** [review]
The fixes look good. Feel free to merge once CI is green
